client/src/main/scala/org/http4s/client/middleware/FollowRedirect.scala

```diff
@@ -41,51 +41,41 @@ object FollowRedirect {
       maxRedirects: Int,
       sensitiveHeaderFilter: CaseInsensitiveString => Boolean = Headers.SensitiveHeaders)(
       client: Client[F])(implicit F: Bracket[F, Throwable]): Client[F] = {
-    def prepareLoop(req: Request[F], redirects: Int): Resource[F, Response[F]] =
-      client.run(req).flatMap { resp =>
-        def redirectUri: Option[Uri] =
-          resp.headers.get(Location).map { loc =>
-            val uri = loc.uri
-            // https://tools.ietf.org/html/rfc7231#section-7.1.2
-            uri.copy(
-              scheme = uri.scheme.orElse(req.uri.scheme),
-              authority = uri.authority.orElse(req.uri.authority),
-              fragment = uri.fragment.orElse(req.uri.fragment)
-            )
-          }
 
-        def dontRedirect: Resource[F, Response[F]] = resp.pure[Resource[F, ?]]
+    def nextRequest(req: Request[F], uri: Uri, method: Method): Request[F] = {
+      // https://tools.ietf.org/html/rfc7231#section-7.1.
+      val nextUri = uri.copy(
+        scheme = uri.scheme.orElse(req.uri.scheme),
+        authority = uri.authority.orElse(req.uri.authority),
+        fragment = uri.fragment.orElse(req.uri.fragment)
+      )
 
-        def stripSensitiveHeaders(nextUri: Uri): Request[F] =
-          if (req.uri.authority != nextUri.authority)
-            req.transformHeaders(_.filterNot(h => sensitiveHeaderFilter(h.name)))
-          else
-            req
+      val req1: Request[F] =
+        if (req.uri.authority != nextUri.authority)
+          req.transformHeaders(_.filterNot(h => sensitiveHeaderFilter(h.name)))
+        else
+          req
 
-        def doRedirect(method: Method): Resource[F, Response[F]] =
-          if (redirects < maxRedirects) {
-            // If we get a redirect response without a location, then there is
-            // nothing to redirect.
-            redirectUri.fold(dontRedirect) { nextUri =>
-              // We can only redirect safely if there is no body or if we've
-              // verified that the body is pure.
-              val nextRequest: Request[F] =
-                stripSensitiveHeaders(nextUri).withMethod(method).withUri(nextUri)
+      val req2: Request[F] = req1.withMethod(method).withUri(nextUri)
+      method match {
+        case GET | HEAD => req2.withEmptyBody
```

Comment:
I find that when I use this numbered idiom, it's easy to add another one and forget the return value.  I think I might prefer the old defs, with descriptive names, to this.  But I like extracting `nextRequest`.

Reply:
Done. Restored some of the old methods.